unifont-9.0.06/man/unigencircles.1
.TH UNIGENCIRCLES 1 "2013 Sep 28"
.SH NAME
unigencircles \- Add dashed combining circles to a unifont.hex file
.SH SYNOPSIS
.br
.B unigencircles
.I combining.txt nonprinting.hex
<
.I unifont.hex
>
.I unifont-circles.hex
.SH DESCRIPTION
.B unigencircles
reads a unifont.hex file from STDIN, adds dashed combining
circles to the hex strings for code points listed in "combining.txt"
but not listed in "nonprinting.hex", and writes the revised set of glyphs
in unifont.hex format to STDOUT.  The resulting combining character glyphs
show the dashed combining circles that appear in The Unicode Standard
code charts.
.PP
For each code point listed in the "combining.txt" file but not listed
in the "nonprinting.hex" file,
.B unigencircles
will superimpose a single-width dashed circle in glyphs that are
single-width (i.e., their hex glyph strings are 32 characters long)
and will superimpose a double-width dashed circle in glyphs that
are double-width (i.e., their hex glyph strings are 64 characters long).
.SH EXAMPLE
.PP
.RS
unigencircles combining.txt nonprinting.hex < unifont.hex > unifont-circles.hex
.RE
.SH FILES
*.hex files for Unifont glyph data
.PP
.B font/ttfsrc/combining.txt
for combining code points
.PP
.B font/plane00/nonprinting.hex
for non-printing code points
.SH SEE ALSO
.BR bdfimplode(1),
.BR hex2bdf(1),
.BR hex2sfd(1),
.BR hexbraille(1),
.BR hexdraw(1),
.BR hexkinya(1),
.BR hexmerge(1),
.BR johab2ucs2(1),
.BR unibdf2hex(1),
.BR unibmp2hex(1),
.BR unicoverage(1),
.BR unidup(1),
.BR unifont(5),
.BR unifont-viewer(1),
.BR unifontchojung(1),
.BR unifontksx(1),
.BR unifontpic(1),
.BR unigenwidth(1),
.BR unihex2bmp(1),
.BR unihex2png(1),
.BR unihexfill(1),
.BR unihexgen(1),
.BR unipagecount(1),
.BR unipng2hex(1)
.SH AUTHOR
.B unigencircles
was written by Paul Hardy.
.SH LICENSE
.B unigencircles
is Copyright \(co 2013 Paul Hardy.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.SH BUGS
No known real bugs exist, except that this software does not perform
extensive error checking on its input files.


unifont-9.0.06/NEWS
2016-12-22: Release 9.0.06

   Refined about 50 Pikto CSUR glyphs.

   Corrected two Osage glyphs and one mathematical operator as noted
   in ChangeLog.

2016-12-10: Release 9.0.05

   Modified Cherokee glyphs to reflect changes in recent Cherokee font
   that the Unicode Consortium adopted for its code charts.  Removed
   serifs to make Cherokee glyphs easier to read.

   Modified Pikto CSUR glyphs after comparison with high-resolution scans
   of Basic Pikto book.

2016-10-29: Release 9.0.04

   Extended "{" and "}" by one pixel so they now come to a point in the center.
   Swapped U+2A05 and U+2A06 so they are now in the correct order.

2016-10-21: Release 9.0.03

   Added CSUR glyphs:
     - Aiha: added doubling combining glyph, omitted from the ConScript
       Unicode Registry (CSUR) code chart.
     - Tonal System: base 16 counting system devised by John William Nystrom
       in the mid-19th century.
     - Pikto: added the Pikto glyphs to Plane 15 Private Use Area, in the
       range specified in the CSUR.

   The giant picture of all of Plane 0 now includes the Plane 0 CSUR glyphs,
   although those are an optional part of Unifont in separate font files.

2016-08-27: Release 9.0.02

   Added 3 more CSUR scripts.  Modifications to some existing glyphs.

2016-07-02: Release 9.0.01

   Updates to support Unicode 9.0.0.

2015-06-28: Release 8.0.01

  Updates to support Unicode 8.0.0.

2014-10-23: Release 7.0.06

   A few new scripts were added:
      U+010350..U+01037F Old Permic
      U+01F650..U+01F67F Ornamental Dingbats
      U+01F780..U+01F7FF Geometrical Shapes Extended
      U+01F800..U+01F8FF Supplemental Arrows-C

   Various changes were made to Makefiles, Perl scripts, and one bash
   shell script to conform to Debian Policy Manual version 3.9.6.

   Several changes were made to Makefiles:
      * Use "set -e" at the start of multi-command shell invocations,
        to trap errors.
      * Use "&&" rather than ";" in multi-command shell invocations,
        to fail on errors.
      * install now uses "-s" flag for C program binaries.

   Other minor changes were made to scripts; see ChangeLog
   for details.

2014-10-17: Release 7.0.05

   Modified ASCII letter 'l' so it would have the same height as
   the other lower-case letters with ascenders; this is one pixel
   taller than the capital letters.  Lowered the horizontal stroke 
   in 't' by one pixel so it would line up with the horizontal
   stroke in 'f'; this is one pixel lower than the x-height of
   the ASCII letters.  Letters elsewhere in Plane 0 that were formed
   with 'k', 'l', or 't' were checked and modified where appropriate
   to conform with the new ASCII glyphs.

   Completed the Miscellaneous Symbols and Pictographs script in Plane 1,
   and modified Plane 1 Emoticons and Transport and Map Symbols scripts.

   Now create file font/compiled/plane01.html, showing a color-coded
   indication of completion of all of Unicode Plane 1 as an HTML
   table in a self-contained HTML file.  That and the compiled/index.html
   are the two HTML files that building the font creates.  The plane01.html
   file gives a color coding to show completion of each 256-glyph block,
   from 0% (pure red) to 100% (pure green).

2014-10-11: Release 7.0.04

   Added five more Supplemental Multilingual Plane (Plane 1) scripts
   and improved others thanks to several contributors.  There are now
   over 5,000 drawn assigned glyphs for Plane 1!

   Reorganized font/plane01 directory.  Several other changes; see
   ChangeLog file for details.

   unihex2sfd and unipagecount now handle all Unicode planes, not just
   Plane 0.


2014-07-01: Release 7.0.03

   Minor release for Debian.  Added a man page for unihexfill(1);
   Debian requires a man page for each program.

2014-06-22: Release 7.0.02

   Fixes a bug in font/Makefile: there were two "done" statements at the
   end of one "for" loop.  Also adds many new Supplemental Multilingual
   Plane (Plane 1) scripts.  See the ChangeLog file for details.

2014-06-21: Release 7.0.01

   Provides a glyph for every visible code point in the Unicode 7.0 Basic
   Mulitlingual Plane.  Also adds more scripts in the Unicode 7.0
   Supplemental Multilingual Plane.  See the ChangeLog file for full
   details.

   Note that the versioning scheme has changed.  Instead of a date, this
   package will use an incrementing minor version number in the future
   ("01", "02", etc.).

2014-02-14: Release 6.3.20140214

   This version adds the GNU Free Documentation License 1.3 to the "COPYING"
   file.

2014-02-04: Release 6.3.20140204

   This version added Qianqian Fang's name to the COPYRIGHT string now
   passed to Fontforge when the TTF fonts are built, for his contributions
   contained in the file "font/plane00/wqy.hex".

   This version also removes "<img>" links to icons on external websites from
   the "hangul/hangul-generation.html" web page, which were retrieved and
   stored locally when the original web page was saved in the "hangul/"
   directory.

2014-02-02: Release 6.3.20140202

   This version is the first designed to easily build multiple fonts.
   To accomplish this, font/Makefile and font/ttfsrc/Makefile were
   modified to pass FONT and COMBINING parameters, giving the names
   of the font file and its associated combining.txt file.

   This version introduces a subset of Unicode's Supplemental Multilingual
   Plane (SMP, or Plane 1).  The file "font/plane01/plane01.hex" contains
   a glyph for every visible code point in the Unicode 5.0 SMP except
   "Cuneiform" and "Cuneiform Numbers and Punctuation"; those two scripts
   require more resolution than what will fit within a 16 by 16 pixel grid.
   Andrew Miller drew well over 1,000 glyphs for this effort, and added
   glyphs for several SMP scripts introduced after Unicode 5.0.

   This version also introduces a subset of the ConScript Unicode
   Registry (CSUR) Private Use Area glyphs.  Wojciech Stryjewski created

unifont-9.0.06/src/unigencircles.c
/*
   unigencircles.c -- program to superimpose dashed combining circles
                      on combining glyphs.

   Author: Paul Hardy, 2013.

   LICENSE:

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 2 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAXSTRING	256


int
main (int argc, char **argv)
{

   char teststring[MAXSTRING];  /* current input line                       */
   int  loc;                    /* Unicode code point of current input line */
   char *gstart;                /* glyph start, pointing into teststring    */

   char combining[0x110000];

   void add_single_circle(char *); /* add a single-width dashed circle */
   void add_double_circle(char *); /* add a double-width dashed circle */

   FILE *infilefp;

   /*
      if (argc != 3) {
         fprintf (stderr,
                "\n\nUsage: %s combining.txt nonprinting.hex < unifont.hex > unifontfull.hex\n\n");
         exit (EXIT_FAILURE);
      }
   */

   /*
      Read the combining characters list.
   */
   /* Start with no combining code points flagged */
   memset (combining, 0, 0x110000 * sizeof (char));

   if ((infilefp = fopen (argv[1],"r")) == NULL) {
      fprintf (stderr,"ERROR - combining characters file %s not found.\n\n",
              argv[1]);
      exit (EXIT_FAILURE);
   }

   /* Flag list of combining characters to add a dashed circle. */
   while (fscanf (infilefp, "%X", &loc) != EOF) combining[loc] = 1;

   fclose (infilefp); /* all done reading combining.txt */

   /* Now read the non-printing glyphs; they never have dashed circles */
   if ((infilefp = fopen (argv[2],"r")) == NULL) {
      fprintf (stderr,"ERROR - nonprinting characters file %s not found.\n\n",
              argv[1]);
      exit (EXIT_FAILURE);
   }

   /* Reset list of nonprinting characters to avoid adding a dashed circle. */
   while (fscanf (infilefp, "%X:%*s", &loc) != EOF) combining[loc] = 0;

   fclose (infilefp); /* all done reading nonprinting.hex */

   /*
      Read the hex glyphs.
   */
   teststring[MAXSTRING - 1] = '\0';   /* so there's no chance we leave array  */
   while (fgets (teststring, MAXSTRING-1, stdin) != NULL) {
      sscanf (teststring, "%X", &loc);     /* loc == the Uniocde code point    */
      gstart = index (teststring,':') + 1; /* start of glyph bitmap            */
      if (combining[loc]) {                /* if a combining character         */
         if (strlen (gstart) < 35) add_single_circle (gstart); /* single-width */
         else add_double_circle (gstart);                      /* double-width */
      }
      printf ("%s", teststring); /* output the new character .hex string */
   }

   exit (EXIT_SUCCESS);
}


/*
   add_single_circle - superimpose a single-width dashed combining circle.
*/
void
add_single_circle (char *glyphstring)
{

   char newstring[256];
   /* Circle hex string pattern is "00000008000024004200240000000000" */
   char circle[32]={0x0,0x0,  /* row  1 */
                    0x0,0x0,  /* row  2 */
                    0x0,0x0,  /* row  3 */
                    0x0,0x0,  /* row  4 */
                    0x0,0x0,  /* row  5 */
                    0x0,0x0,  /* row  6 */
                    0x2,0x4,  /* row  7 */
                    0x0,0x0,  /* row  8 */
                    0x4,0x2,  /* row  9 */
                    0x0,0x0,  /* row 10 */
                    0x2,0x4,  /* row 11 */
                    0x0,0x0,  /* row 12 */
                    0x0,0x0,  /* row 13 */
                    0x0,0x0,  /* row 14 */
                    0x0,0x0,  /* row 15 */
                    0x0,0x0}; /* row 16 */

   int digit1, digit2; /* corresponding digits in each string */

   int i; /* index variables */

   /* for each character position, OR the corresponding circle glyph value */
   for (i = 0; i < 32; i++) {
      glyphstring[i] = toupper (glyphstring[i]);

      /* Convert ASCII character to a hexadecimal integer */
      digit1 = (glyphstring[i] <= '9') ?
               (glyphstring[i] - '0') : (glyphstring[i] - 'A' + 0xA);

      /* Superimpose dashed circle */
      digit2 = digit1 | circle[i];

      /* Convert hexadecimal integer to an ASCII character */
      newstring[i] = (digit2 <= 9) ?
                     ('0' + digit2) : ('A' + digit2 - 0xA);
   }

   /* Terminate string for output */
   newstring[i++] = '\n';
   newstring[i++] = '\0';

   strncpy (glyphstring, newstring, i);

   return;
}


/*
   add_double_circle - superimpose a single-width dashed combining circle.
*/
void
add_double_circle (char *glyphstring)
{

   char newstring[256];
   /* Circle hex string pattern is "00000008000024004200240000000000" */
   char circle[64]={0x0,0x0,0x0,0x0,  /* row  1 */
                    0x0,0x0,0x0,0x0,  /* row  2 */
                    0x0,0x0,0x0,0x0,  /* row  3 */
                    0x0,0x0,0x0,0x0,  /* row  4 */
                    0x0,0x0,0x0,0x0,  /* row  5 */
                    0x0,0x0,0x0,0x0,  /* row  6 */
                    0x0,0x2,0x4,0x0,  /* row  7 */
                    0x0,0x0,0x0,0x0,  /* row  8 */
                    0x0,0x4,0x2,0x0,  /* row  9 */
                    0x0,0x0,0x0,0x0,  /* row 10 */
                    0x0,0x2,0x4,0x0,  /* row 11 */
                    0x0,0x0,0x0,0x0,  /* row 12 */
                    0x0,0x0,0x0,0x0,  /* row 13 */
                    0x0,0x0,0x0,0x0,  /* row 14 */
                    0x0,0x0,0x0,0x0,  /* row 15 */
                    0x0,0x0,0x0,0x0}; /* row 16 */

   int digit1, digit2; /* corresponding digits in each string */

   int i; /* index variables */

   /* for each character position, OR the corresponding circle glyph value */
   for (i = 0; i < 64; i++) {
      glyphstring[i] = toupper (glyphstring[i]);

      /* Convert ASCII character to a hexadecimal integer */
      digit1 = (glyphstring[i] <= '9') ?
               (glyphstring[i] - '0') : (glyphstring[i] - 'A' + 0xA);

      /* Superimpose dashed circle */
      digit2 = digit1 | circle[i];

      /* Convert hexadecimal integer to an ASCII character */
      newstring[i] = (digit2 <= 9) ?
                     ('0' + digit2) : ('A' + digit2 - 0xA);
   }

   /* Terminate string for output */
   newstring[i++] = '\n';
   newstring[i++] = '\0';

   strncpy (glyphstring, newstring, i);

   return;
}

